### PR TITLE
Create a Generic Build step which can send Build Information

### PR DESCRIPTION
**Pull request by rain-on**
This makes use of the early-beta Octopus java sdk, and sends build information to an Octopus server _without_ requiring Octo.exe.

Note: this is using unstable versions for the OctopusSDK - which does not support the 'overwrite' field on the buildinformation creation endpoint.

### CODE REVIEW

File: octopus-agent/src/main/java/octopus/teamcity/agent/buildinformation/BuildVcsData.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+  @Override
+  public List<Commit> getCommits() {
+    final List<Change> changes = buildDataFetcher.fetchChanges();
+
+    final List<Commit> commits = new ArrayList<>();
+    for (Change change : changes) {
+
+      final Commit c = new Commit(change.getVersion(), change.getComment());
+      commits.add(c);
```
Comment (long-sean) [inline octopus-agent/src/main/java/octopus/teamcity/agent/buildinformation/BuildVcsData.java:58]: simplify to `commits.add(new Commit....)` ?

File: octopus-agent/src/main/java/octopus/teamcity/agent/generic/in_sdk/ConnectDataBuilder.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  public ConnectDataBuilder withConnectTimeout(final Duration connectTimeout) {
+    this.connectTimeout = connectTimeout;
+    return this;
+  }
+
+  public ConnectData build() {
+    Preconditions.checkNotNull(octopusServerUrl, "Server URL cannot be null");
+    Preconditions.checkNotNull(apiKey, "Api Key cannot be null");
```
Comment (long-sean) [inline octopus-agent/src/main/java/octopus/teamcity/agent/generic/in_sdk/ConnectDataBuilder.java:52]: Case consistency in messages

File: octopus-agent/src/main/java/octopus/teamcity/agent/pushpackage/FileSelector.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.List;
+
+import com.google.common.collect.Lists;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class FileSelector {
+  private static final Logger LOG = LogManager.getLogger();
+
+  public static List<File> getMatchingFiles(final Path rootPath, final List<String> globs) {
```
Comment (long-sean) [inline octopus-agent/src/main/java/octopus/teamcity/agent/pushpackage/FileSelector.java:36]: I could not find any validation in the chain that ensures globs is not null at this point. Is parameter validation done before start() and I missed it?

File: octopus-agent/src/test/java/octopus/teamcity/agent/pushpackage/PushPackageUploaderTest.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+package octopus.teamcity.agent.pushpackage;
+
+import org.junit.jupiter.api.Test;
+
+class PushPackageUploaderTest {
+
+  // final OctopusDeployServer server = null;
+
+  @Test
+  public void doesThisWorkInMyArea() {
```
Comment (long-sean) [inline octopus-agent/src/test/java/octopus/teamcity/agent/pushpackage/PushPackageUploaderTest.java:25]: Remove or complete?

File: octopus-server/src/main/resources/buildServerResources/v2/subpages/viewPushPackageParameters.jsp [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    Package paths:
+    <strong><props:displayValue name="${keys.packagePathsKey}" emptyValue="not specified"/></strong>
+</div>
+
+<div class="parameter">
+    Overwrite Mode:
+    <strong><props:displayValue name="${keys.overwriteModeKey}" emptyValue="not specified"/></strong>
+</div>
+
+<%--
```
Comment (long-sean) [inline octopus-server/src/main/resources/buildServerResources/v2/subpages/viewPushPackageParameters.jsp:36]: Does this need to be in or out?

**Review by long-sean — APPROVED** [review]
Some "nit-picking" comments, up to you if you address

File: octopus-agent/src/main/java/octopus/teamcity/agent/buildinformation/BuildVcsData.java (hunk 1 above)
Comment (rain-on) [reply octopus-agent/src/main/java/octopus/teamcity/agent/buildinformation/BuildVcsData.java:58]: done

File: octopus-agent/src/main/java/octopus/teamcity/agent/generic/in_sdk/ConnectDataBuilder.java (hunk 2 above)
Comment (rain-on) [reply octopus-agent/src/main/java/octopus/teamcity/agent/generic/in_sdk/ConnectDataBuilder.java:52]: exists in SDK - will probably need fixing there.

**Comment by rain-on** [comment]
Overtaken by other PRs.

**Pull request closed without merging**